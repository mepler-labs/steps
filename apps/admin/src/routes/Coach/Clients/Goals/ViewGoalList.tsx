import { addAlert } from 'actions/alerts';
import { getClients } from 'actions/clients';
import Button from 'atoms/Buttons/Button';
import { Alert, AlertLevel } from 'components/Alert/types';
import GoalList from 'components/Goals/GoalList';
import NoGoals from 'components/Goals/NoGoals';
import { Box, Flex } from 'grid-styled';
import React from 'react';
import { connect } from 'react-redux';
import { Link, History } from 'react-router-dom';
import { Client } from 'reducers/clients';
import { bindActionCreators } from 'redux';
import Goal from 'components/Goals/Goal';
import styled from 'styled-components';

interface Props {
  actions: {
    addAlert(alert: Alert): void;
    getClients(): Promise<void>;
  };
  client?: Client;
  history: History;
}

interface GoalListLayoutProps {
  goals: string[];
  client: Client;
  history: History;
}

export const GoalListLayout: React.SFC<GoalListLayoutProps> = ({
  goals,
  client,
  history,
}) => (
  <div>
    <GoalList goals={goals}>
      {childProps => (
        <Goal
          {...childProps}
          onEdit={() =>
            history.push(`/clients/${client.id}/goals/${childProps.index}/edit`)
          }
        />
      )}
    </GoalList>
    <Flex justifyContent="center">
      <Link to={`/clients/${client.id}/goals/new`}>
        <Button>New Goal</Button>
      </Link>
    </Flex>
  </div>
);

export class ViewGoalList extends React.Component<Props> {
  componentWillMount() {
    this.props.actions.getClients().catch(err => {
      this.props.actions.addAlert({
        level: AlertLevel.Error,
        message: err.message,
        id: 'view-goal-get-clients-error',
      });
    });
  }

  render() {
    const { client, history } = this.props;
    return (
      <Box width={1} p={4}>
        {this.hasGoals() ? (
          <GoalListLayout
            goals={client ? client.goals : []}
            client={client}
            history={history}
          />
        ) : (
          <NoGoals client={this.props.client} />
        )}
      </Box>
    );
  }

  private hasGoals(): boolean {
    return (
      this.props.client &&
      this.props.client.goals &&
      this.props.client.goals.length > 0
    );
  }
}

const mapStateToProps = (state, props) => {
  return {
    client: state.clients.clients.find(c => c.id == props.match.params.id),
  };
};

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ addAlert, getClients }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ViewGoalList);
