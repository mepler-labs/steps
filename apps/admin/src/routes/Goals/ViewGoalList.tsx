import GoalList from 'components/Goals/GoalList';
import React from 'react';
import NoGoals from 'components/Goals/NoGoals';
import { Box } from 'grid-styled';
import { Link } from 'react-router-dom';
import { Goal as GoalType } from 'components/Goals/types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getGoals } from 'actions/goals';

interface Props {
  actions: {
    getGoals(): GoalType[];
  };
  goals: GoalType[];
}

export const GoalListLayout: React.SFC<{ goals: GoalType[] }> = ({ goals }) => (
  <div>
    <GoalList goals={goals} />
    <Link to="/clients/1/goals/new">New Goal</Link>
  </div>
);

export class ViewGoalList extends React.Component<Props> {
  componentWillMount() {
    this.props.actions.getGoals();
  }

  render() {
    return (
      <Box width={1} p={4}>
        {this.props.goals && this.props.goals.length > 0 ? (
          <GoalListLayout goals={this.props.goals} />
        ) : (
          <NoGoals />
        )}
      </Box>
    );
  }
}
const mapStateToProps = ({ goals }) => {
  return {
    goals,
  };
};

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getGoals }, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ViewGoalList);
