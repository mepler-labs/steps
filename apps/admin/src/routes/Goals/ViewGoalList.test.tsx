import { shallow } from 'enzyme';
import React from 'react';
import { GoalListLayout, ViewGoalList } from './ViewGoalList';

describe('ViewGoalList.tsx', () => {
  it('renders correctly', () => {
    const actions = {
      getGoals: jest.fn().mockReturnValue(Promise.resolve()),
      addAlert: jest.fn(),
    };
    const goals = [
      {
        text: 'first goal',
      },
      {
        text: 'second goal',
      },
    ];

    const wrapper = shallow(<ViewGoalList goals={goals} actions={actions} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('loads goals on mount', () => {
    const actions = {
      getGoals: jest.fn().mockReturnValue(Promise.resolve()),
      addAlert: jest.fn(),
    };

    const wrapper = shallow(<ViewGoalList goals={[]} actions={actions} />);

    expect(actions.getGoals).toHaveBeenCalled();
  });

  it('displays a list of goals if there are goals', () => {
    const actions = {
      getGoals: jest.fn().mockReturnValue(Promise.resolve()),
      addAlert: jest.fn(),
    };

    const goals = [{ text: 'first goal' }, { text: 'second goal' }];
    const wrapper = shallow(<ViewGoalList goals={goals} actions={actions} />);
    const goalList = wrapper.find(GoalListLayout);

    expect(goalList).toHaveLength(1);
  });

  it('displays an error message if load goals fails', done => {
    const actions = {
      getGoals: jest
        .fn()
        .mockReturnValue(Promise.reject({ message: 'some error' })),
      addAlert: jest.fn(),
    };

    const wrapper = shallow(<ViewGoalList goals={[]} actions={actions} />);

    process.nextTick(() => {
      expect(actions.addAlert).toHaveBeenCalledWith({
        level: 'error',
        message: 'some error',
        id: 'view-goal-error',
      });
      done();
    });
  });
});
