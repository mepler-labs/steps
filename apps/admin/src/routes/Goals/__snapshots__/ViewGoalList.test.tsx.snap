// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewGoalList.tsx renders correctly 1`] = `
<Box
  p={4}
  theme={
    Object {
      "breakpoints": Array [
        "40em",
        "52em",
        "64em",
      ],
      "fontSizes": Array [
        12,
        14,
        16,
        20,
        24,
        32,
        48,
        64,
        96,
        128,
      ],
      "space": Array [
        0,
        4,
        8,
        16,
        32,
        64,
        128,
        256,
        512,
      ],
    }
  }
  width={1}
>
  <Component
    goals={
      Array [
        Object {
          "text": "first goal",
        },
        Object {
          "text": "second goal",
        },
      ]
    }
  />
</Box>
`;
